src/rlang/dyn-array.c

```diff
@@ -0,0 +1,71 @@
+#include <rlang.h>
+#include "dyn-array.h"
+
+#define R_DYN_ARRAY_GROWTH_FACTOR 2
+
+static
+sexp* attribs_dyn_array = NULL;
+
+
+struct r_dyn_array* r_new_dyn_array(r_ssize capacity,
+                                    r_ssize elt_byte_size) {
+  sexp* shelter = KEEP(r_new_list(2));
+  r_poke_attrib(shelter, attribs_dyn_array);
+  r_mark_object(shelter);
+
+  sexp* vec_raw = r_new_raw(sizeof(struct r_dyn_array));
+  r_list_poke(shelter, 0, vec_raw);
+
+  sexp* arr_raw = r_new_raw(r_ssize_mult(capacity, elt_byte_size));
+  r_list_poke(shelter, 1, arr_raw);
+
+  struct r_dyn_array* p_arr = r_raw_deref(vec_raw);
+  p_arr->shelter = shelter;
+  p_arr->count = 0;
+  p_arr->capacity = capacity;
+  p_arr->growth_factor = R_DYN_ARRAY_GROWTH_FACTOR;
+  p_arr->v_data = r_raw_deref(arr_raw);
+  p_arr->elt_byte_size = elt_byte_size;
+
+  FREE(1);
+  return p_arr;
+}
+
+void r_arr_push_back(struct r_dyn_array* p_arr, void* p_elt) {
+  r_ssize count = ++p_arr->count;
+  r_arr_grow(p_arr, count);
```

Comment:
There may be some performance advantage to inlining the condition of the most common case to avoid a function call:

```c
if (capacity <= p_arr->capacity) {
    return;
}
```

It made a difference when working on vec-order, which is the only reason I bring it up

Reply:
Given this and your other comment, I've now inlined `r_arr_grow()` inside `r_arr_push_back()` since it doesn't seem that useful: https://github.com/r-lib/rlang/commit/2aab3e12e2982d687208e312c018afbf44908852